File: code/python/ch22/practice/warmup_04_next_greater_element.py
"""
Warmup 4: Next Greater Element
==================================
Chapter 22: Stacks & Queues — Order Matters

PROBLEM
-------
Given an array of integers, for each element find the next greater element
to its right. If no greater element exists, use -1.

CONSTRAINTS
-----------
- 0 <= len(arr) <= 10^5
- -10^9 <= arr[i] <= 10^9

EXAMPLES
--------
Input: 4 5 2 10 8
Output: [5, 10, 10, -1, -1]

HINT
----
Use a monotonic stack: process from right to left, maintaining a decreasing stack.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
"""


def solve(arr: list[int]) -> list[int]:
    """Return list of next greater elements for each position."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    if line:
        arr = list(map(int, line.split()))
    else:
        arr = []
    print(solve(arr))

File: code/cpp/ch22/practice/warmup_04_next_greater_element.cpp
/*
 * Warmup 4: Next Greater Element
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * PROBLEM: For each element, find the next greater element to its right (-1 if none).
 * EXAMPLES: solve({4,5,2,10,8}) -> {5,10,10,-1,-1}
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

vector<int> solve(vector<int> arr) {
    // TODO: Replace this with your solution
    return vector<int>(arr.size(), -1);
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    vector<int> result = solve(arr);
    for (int x : result) cout << x << " ";
    cout << endl;
    return 0;
}
